File: ArrayMostFreq.java
import java.util.HashMap;
import java.util.Map;
public class ArrayMostFreq {
    public static void main(String[] args) {
        // NOTE: The following input values are used for testing your solution.

        // mostFrequent(array1) should return 1.
        int[] array1 = {1, 3, 1, 3, 2, 1};
        if (ArrayMostFreq.mostFreqent(array1) == 1) {
            System.out.printf("Correct! ArrayMostFreq.mostFreqent(array1)=%d", ArrayMostFreq.mostFreqent(array1) );
        }
        // mostFrequent(array2) should return 3.
        int[] array2 = {3, 3, 1, 3, 2, 1};
        // mostFrequent(array3) should return null.
        int[] array3 = {};
        // mostFrequent(array4) should return 0.
        int[] array4 = {0};
        // mostFrequent(array5) should return -1.
        int[] array5 = {0, -1, 10, 10, -1, 10, -1, -1, -1, 1};
    }
    // Implement your solution below.
    public static Integer mostFreqent(int[] arr) {
        Integer maxItem = null;
        Integer max = 0;
        HashMap<Integer, Integer> histogram = new HashMap<Integer, Integer>();
        for (int num : arr) {
            histogram.compute(num, (k, value) -> value == null ? 0 : value + 1);
            // histogram.put(num, histogram.getOrDefault(num, 0) + 1);
        }
        for (Map.Entry<Integer, Integer> e : histogram.entrySet()) {
            if (e.getValue() >= max) {
                max = e.getValue();
                maxItem = e.getKey();
            };
        };
        return maxItem;
    }
}

File: arrayMostFreq.py
# Implement your function below.
import unittest

'''
O(n)
'''
def most_frequent(arr):
    if (len(arr) == 0):
        return None
    elif len(arr) == 1:
        return arr.pop()
    max = 0
    result = None
    hist = dict()
    for num in arr:
        hist[num] = hist.get(num, 0) + 1
    for num, fq in hist.items():
        if max < fq:
            max = fq
            result = num
    return result

# NOTE: The following input values will be used for testing your solution.
# most_frequent(list1) should return 1
list1 = [1, 3, 1, 3, 2, 1]
# most_frequent(list2) should return 3
list2 = [3, 3, 1, 3, 2, 1]
# most_frequent(list3) should return None
list3 = []
# most_frequent(list4) should return 0
list4 = [0]
# most_frequent(list5) should return -1
list5 = [0, -1, 10, 10, -1, 10, -1, -1, -1, 1]

class TestMe(unittest.TestCase):
    def test_1(self):
        self.assertEqual(most_frequent(list1), 1)
    def test_2(self):
        self.assertEqual(most_frequent(list2), 3)
    def test_3(self):
        self.assertEqual(most_frequent(list3), None)
    def test_4(self):
        self.assertEqual(most_frequent(list4), 0)
    def test_5(self):
        self.assertEqual(most_frequent(list5), -1)

if __name__ == "__main__":
    unittest.main()